nemo/utils/config_utils.py

```diff
@@ -0,0 +1,170 @@
+    1)  If the sub-config key exists in the `update_cfg`, but does not exist in ModelPT config:
+        - Add the sub-config from update_cfg to ModelPT config
+
+    2) If the sub-config key does not exist in `update_cfg`, but exists in ModelPT config:
+        - Remove the sub-config from the ModelPT config; iff the `drop_missing_subconfigs` flag is set.
+
+    Args:
+        model_cfg: A DictConfig instantiated from the ModelPTConfig subclass.
+        update_cfg: A DictConfig that mirrors the structure of `model_cfg`, used to update its default values.
+        subconfig_key: A str key used to check and update the sub-config.
+        drop_missing_subconfigs: A bool flag, whether to allow deletion of the ModelPTConfig sub-config,
+            if its mirror sub-config does not exist in the `update_cfg`.
+
+    Returns:
+        The updated DictConfig for the ModelPTConfig
+    """
+    with open_dict(model_cfg.model):
+        # If update config has the key, but model cfg doesnt have the key
+        # Add the update cfg subconfig to the model cfg
+        if subconfig_key in update_cfg.model and subconfig_key not in model_cfg.model:
+            model_cfg.model[subconfig_key] = update_cfg.model[subconfig_key]
+
+        # If update config does not the key, but model cfg has the key
+        # Remove the model cfg subconfig in order to match layout of update cfg
+        if subconfig_key not in update_cfg.model and subconfig_key in model_cfg.model:
+            if drop_missing_subconfigs:
+                model_cfg.model.pop(subconfig_key)
+
+    return model_cfg
+
+
+def _add_subconfig_keys(model_cfg: DictConfig, update_cfg: DictConfig, subconfig_key: str):
+    """
+    For certain sub-configs, the default values specified by the ModelPTConfig class is insufficient.
+    In order to support every potential value in the merge between the `update_cfg`, it would require
+    explicit definition of all possible cases.
+
+    An example of such a case is Optimizers, and their equivalent Schedulers. All optimizers share a few basic
+    details - such as name and lr, but almost all require additional parameters - such as weight decay.
+    It is impractical to create a config for every single optimizer + every single scheduler combination.
+
+    In such a case, we perform a dual merge. The Optim and Sched Dataclass contain the bare minimum essential
+    components. The extra values are provided via update_cfg.
+
+    In order to enable the merge, we first need to update the update sub-config to incorporate the keys,
+    with dummy temporary values (merge update config with model config). This is done on a copy of the
+    update sub-config, as the actual override values might be overriden by the ModelPTConfig defaults.
+
+    Then we perform a merge of this temporary sub-config with the actual override config in a later step
+    (merge model_cfg with original update_cfg, done outside this function).
```

Comment:
I don't think this is a correct use of dataclasses. For the optimizer, you should only have the bare minimum that is the same across all optimizers parameters. If there is nothing, then it should just be empty. In which case, omegaconf.merge is more than sufficient.

Reply:
Nvm